### PR TITLE
OKTA-509021 - Enroll password authenticator v3 parity spec

### PR DESCRIPTION
**Pull request by vijetmahabaleshwar-okta**
## Description:
- This PR adds v3 parity to the Enroll Password Authenticator Testcafe spec

## PR Checklist

- [ ] Have you verified the basic functionality for this change?
- [ ] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [ ] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-509021](https://oktainc.atlassian.net/browse/OKTA-509021)

### CODE REVIEW

File: test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js [hunk 1]
```diff
@@ -98,8 +98,8 @@ test.requestHooks(successMock)('should have the correct reqiurements', async t =
   await t.expect(enrollPasswordPage.getRequirements()).contains('A symbol');
   await t.expect(enrollPasswordPage.getRequirements()).contains('Does not include your first name');
   await t.expect(enrollPasswordPage.getRequirements()).contains('Does not include your last name');
-  await t.expect(enrollPasswordPage.getRequirements()).contains('At least 2 hour(s) must have elapsed since you last changed your password');
+  // await t.expect(enrollPasswordPage.getRequirements()).contains('At least 2 hour(s) must have elapsed since you last changed your password');
```
Comment (vijetmahabaleshwar-okta) [inline test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js:101]: These 2 messages aren't available in the v3 widget for password authenticator enrollment page.
I'm not sure if these messages are from server through the response or rendered by the widget.

File: test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js [hunk 2]
```diff
@@ -59,7 +61,6 @@ test.requestHooks(successMock)('should have both password and confirmPassword fi
   await enrollPasswordPage.waitForErrorBox();
   await t.expect(enrollPasswordPage.hasPasswordError()).eql(false);
   await t.expect(enrollPasswordPage.getConfirmPasswordError()).eql('New passwords must match');
-  await t.expect(enrollPasswordPage.getErrorBoxText()).eql('We found some errors. Please review the form and make corrections.');
```
Comment (vijetmahabaleshwar-okta) [inline test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js:62]: This was a redundant assertion.
`await enrollPasswordPage.waitForErrorBox();` does the same check

File: test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js (hunk 1 above)
Comment (glenfannin-okta) [reply test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js:101]: In v3, UX & Product actually made a decision to _not_ display the server side messages because they can be sort of confusing. There is another ticket pending final work from UX that will re-introduce these server side messages but in a different context.

For now, I would actually recommend wrapping this in a condition (to not impact v2 tests but exclude it for v3), this is one of the rare times that we include conditional statements in the test, you can mimic what we did in this test: https://github.com/okta/okta-signin-widget/blob/0.1/test/testcafe/spec/ReEnrollAuthenticatorWarningPasswordView_spec.js#L84-L89

File: test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js [hunk 3]
```diff
@@ -98,8 +98,8 @@ test.requestHooks(successMock)('should have the correct reqiurements', async t =
   await t.expect(enrollPasswordPage.getRequirements()).contains('A symbol');
   await t.expect(enrollPasswordPage.getRequirements()).contains('Does not include your first name');
   await t.expect(enrollPasswordPage.getRequirements()).contains('Does not include your last name');
-  await t.expect(enrollPasswordPage.getRequirements()).contains('At least 2 hour(s) must have elapsed since you last changed your password');
+  // await t.expect(enrollPasswordPage.getRequirements()).contains('At least 2 hour(s) must have elapsed since you last changed your password');
   await t.expect(enrollPasswordPage.getRequirements()).contains('No parts of your username');
-  await t.expect(enrollPasswordPage.getRequirements()).contains('Your password cannot be any of your last 4 passwords');
+  // await t.expect(enrollPasswordPage.getRequirements()).contains('Your password cannot be any of your last 4 passwords');
```
Comment (glenfannin-okta) [inline test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js:103]: You can group this and the one above in the same condition.

File: test/testcafe/framework/page-objects/components/BaseFormObject.js [hunk 4]
```diff
@@ -176,10 +176,11 @@ export default class BaseFormObject {
     await this.t.click(this.el.find(CANCEL_BUTTON_SELECTOR));
   }
 
-  getSaveButtonLabel() {
+  getSaveButtonLabel(name = 'Next') {
     // in v3 buttons dont have a value prop
     if (userVariables.v3) {
-      return within(this.el).getByRole('button').textContent;
+      const submitButton = this.getButton(name);
+      return submitButton.textContent;
```
Comment (glenfannin-okta) [inline test/testcafe/framework/page-objects/components/BaseFormObject.js:183]: Is this change needed? Since there should only be one button on the page and that button should be the "Next" button?

File: playground/mocks/data/idp/idx/authenticator-enroll-password.json [hunk 5]
```diff
@@ -21,7 +21,8 @@
                 {
                   "name": "passcode",
                   "label": "Enter password",
-                  "secret": true
+                  "secret": true,
+                  "required": true
```
Comment (shawyu-okta) [inline playground/mocks/data/idp/idx/authenticator-enroll-password.json:25]: Can you double-check to ensure the `required` field appears here rather than a level up between line 18 and 19? I have seen it appear at both levels.

File: test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js (hunk 1 above)
Comment (shawyu-okta) [reply test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js:101]: Do we have a ticket tracking the re-introduction of this message with new UX? If so, we should include it in the comment explaining why we are skipping this assertion in v3.

File: test/testcafe/framework/page-objects/components/BaseFormObject.js (hunk 4 above)
Comment (shawyu-okta) [reply test/testcafe/framework/page-objects/components/BaseFormObject.js:183]: I do think in some situations there is more than one button on the page -- is that why you added the query for a specific button by its name @vijetmahabaleshwar-okta?

File: playground/mocks/data/idp/idx/authenticator-enroll-password.json (hunk 5 above)
Comment (glenfannin-okta) [reply playground/mocks/data/idp/idx/authenticator-enroll-password.json:25]: Oh yea, looks like it should technically be at the `credentials` level. 

image

File: test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js (hunk 1 above)
Comment (glenfannin-okta) [reply test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js:101]: Yep! We have that ticket documented in the other test too. `OKTA-533383`

If you want to copy the exact comment that the link above has I think that is totally acceptable since they are pretty much equivalent.

File: test/testcafe/framework/page-objects/components/BaseFormObject.js (hunk 4 above)
Comment (vijetmahabaleshwar-okta) [reply test/testcafe/framework/page-objects/components/BaseFormObject.js:183]: That's right @shawyu-okta. 
On the password enrollment page, the "show/hide" password is a button. And the test fails since there are multiple buttons that match the selector. Had to do this change to get over that issue. @glenfannin-okta

**Comment by vijetmahabaleshwar-okta** [comment]
Thanks for the comments, @shawyu-okta and @glenfannin-okta.
I have addressed your concerns in the new commit. Please review when possible.

**Review by glenfannin-okta — APPROVED** [review]

File: test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js [hunk 6]
```diff
@@ -18,11 +18,13 @@ const errorMock = RequestMock()
...
   .respond(xhrAuthenticatorEnrollPasswordError, 403);
 
-fixture('Authenticator Enroll Password');
+fixture('Authenticator Enroll Password')
+  .meta('v3', true);
 
 async function setup(t) {
   const enrollPasswordPage = new FactorEnrollPasswordPageObject(t);
   await enrollPasswordPage.navigateToPage();
+  await t.expect(Selector('form').exists).eql(true);
```
Comment (lesterchoi-okta) [inline test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js:27]: I think we do this a lot for v3. Feel free to move it into a base object if/when it's re-used.

**Review by lesterchoi-okta — APPROVED** [review]

File: test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js (hunk 6 above)
Comment (shawyu-okta) [reply test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js:27]: @leemchale-okta tried to add this at the baseform level and found some tests failed since they don't actually show the form at all. Perhaps we can look at a more nuanced inclusion (e.g. flag in the constructor or utility method for setup in the baseform)

**Review by shawyu-okta — APPROVED** [review]
